File: packages/contracts/src/schemas/external-adapter.ts
import { Type, type Static } from "@sinclair/typebox";
import {
  JSON_SCHEMA_DRAFT_2020_12,
  RUNX_CONTRACT_IDS,
  RUNX_LOGICAL_SCHEMAS,
  type DeepReadonly,
  dateTimeStringSchema,
  stringEnum,
  unknownRecordSchema,
  validateContractSchema,
} from "../internal.js";
import { resolutionRequestSchema } from "./resolution.js";
import { referenceSchema } from "./spine.js";

export const externalAdapterProtocolVersion = "runx.external_adapter.v1" as const;

// Process transport v1 is deliberately small: the Rust supervisor writes one
// invocation JSON document to stdin, accepts exactly one response JSON document
// on stdout, and treats stderr as diagnostic text only.
const externalAdapterTransports = ["process", "http"] as const;
const externalAdapterStatuses = [
  "completed",
  "failed",
  "host_resolution_requested",
  "cancelled",
] as const;
const externalAdapterCredentialPurposes = [
  "provider_api",
  "registry",
  "artifact_store",
  "webhook_verification",
] as const;

const nonEmptyStringArraySchema = Type.Array(Type.String({ minLength: 1 }), { minItems: 1 });

export const externalAdapterTransportSchema = Type.Object(
  {
    kind: stringEnum(externalAdapterTransports),
    command: Type.Optional(Type.String({ minLength: 1 })),
    args: Type.Optional(Type.Array(Type.String())),
    endpoint: Type.Optional(Type.String({ minLength: 1 })),
  },
  { additionalProperties: false },
);

export type ExternalAdapterTransportContract = DeepReadonly<Static<typeof externalAdapterTransportSchema>>;

export const externalAdapterCredentialNeedSchema = Type.Object(
  {
    purpose: stringEnum(externalAdapterCredentialPurposes),
    provider: Type.String({ minLength: 1 }),
    scope_refs: Type.Optional(Type.Array(referenceSchema)),
    required: Type.Boolean(),
  },
  { additionalProperties: false },
);

export type ExternalAdapterCredentialNeedContract =
  DeepReadonly<Static<typeof externalAdapterCredentialNeedSchema>>;

export const externalAdapterSandboxIntentSchema = Type.Object(
  {
    profile: Type.String({ minLength: 1 }),
    network: Type.Boolean(),
    cwd_policy: Type.String({ minLength: 1 }),
    writable_paths: Type.Optional(Type.Array(Type.String({ minLength: 1 }))),
  },
  { additionalProperties: false },
);

export type ExternalAdapterSandboxIntentContract =
  DeepReadonly<Static<typeof externalAdapterSandboxIntentSchema>>;

export const externalAdapterTimeoutsSchema = Type.Object(
  {
    startup_ms: Type.Integer({ minimum: 1 }),
    invocation_ms: Type.Integer({ minimum: 1 }),
  },
  { additionalProperties: false },
);

export type ExternalAdapterTimeoutsContract = DeepReadonly<Static<typeof externalAdapterTimeoutsSchema>>;

export const externalAdapterManifestV1Schema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.externalAdapterManifest),
    protocol_version: Type.Literal(externalAdapterProtocolVersion),
    adapter_id: Type.String({ minLength: 1 }),
    name: Type.String({ minLength: 1 }),
    version: Type.String({ minLength: 1 }),
    supported_source_types: nonEmptyStringArraySchema,
    transport: externalAdapterTransportSchema,
    timeouts: externalAdapterTimeoutsSchema,
    credential_needs: Type.Optional(Type.Array(externalAdapterCredentialNeedSchema)),
    sandbox_intent: externalAdapterSandboxIntentSchema,
    metadata: Type.Optional(unknownRecordSchema()),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.externalAdapterManifest,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.externalAdapterManifest,
    additionalProperties: false,
  },
);

export type ExternalAdapterManifestContract =
  DeepReadonly<Static<typeof externalAdapterManifestV1Schema>>;

export const externalAdapterCredentialReferenceSchema = Type.Object(
  {
    credential_ref: referenceSchema,
    provider: Type.String({ minLength: 1 }),
    purpose: stringEnum(externalAdapterCredentialPurposes),
  },
  { additionalProperties: false },
);

export type ExternalAdapterCredentialReferenceContract =
  DeepReadonly<Static<typeof externalAdapterCredentialReferenceSchema>>;

export const externalAdapterCredentialRequestV1Schema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.externalAdapterCredentialRequest),
    protocol_version: Type.Literal(externalAdapterProtocolVersion),
    request_id: Type.String({ minLength: 1 }),
    adapter_id: Type.String({ minLength: 1 }),
    invocation_id: Type.String({ minLength: 1 }),
    credential_refs: Type.Array(externalAdapterCredentialReferenceSchema, { minItems: 1 }),
    requested_at: dateTimeStringSchema(),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.externalAdapterCredentialRequest,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.externalAdapterCredentialRequest,
    additionalProperties: false,
  },
);

export type ExternalAdapterCredentialRequestContract =
  DeepReadonly<Static<typeof externalAdapterCredentialRequestV1Schema>>;

export const externalAdapterInvocationV1Schema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.externalAdapterInvocation),
    protocol_version: Type.Literal(externalAdapterProtocolVersion),
    invocation_id: Type.String({ minLength: 1 }),
    adapter_id: Type.String({ minLength: 1 }),
    run_id: Type.String({ minLength: 1 }),
    step_id: Type.String({ minLength: 1 }),
    source_type: Type.String({ minLength: 1 }),
    skill_ref: Type.String({ minLength: 1 }),
    harness_ref: referenceSchema,
    host_ref: referenceSchema,
    inputs: unknownRecordSchema(),
    resolved_inputs: Type.Optional(unknownRecordSchema()),
    cwd: Type.Optional(Type.String({ minLength: 1 })),
    receipt_dir: Type.Optional(Type.String({ minLength: 1 })),
    env: Type.Optional(unknownRecordSchema()),
    credential_refs: Type.Optional(Type.Array(externalAdapterCredentialReferenceSchema)),
    metadata: Type.Optional(unknownRecordSchema()),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.externalAdapterInvocation,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.externalAdapterInvocation,
    additionalProperties: false,
  },
);

export type ExternalAdapterInvocationContract =
  DeepReadonly<Static<typeof externalAdapterInvocationV1Schema>>;

export const externalAdapterArtifactObservationSchema = Type.Object(
  {
    artifact_ref: referenceSchema,
    summary: Type.Optional(Type.String({ minLength: 1 })),
  },
  { additionalProperties: false },
);

export type ExternalAdapterArtifactObservationContract =
  DeepReadonly<Static<typeof externalAdapterArtifactObservationSchema>>;

export const externalAdapterErrorObservationSchema = Type.Object(
  {
    code: Type.String({ minLength: 1 }),
    message: Type.String({ minLength: 1 }),
    retryable: Type.Boolean(),
  },
  { additionalProperties: false },
);

export type ExternalAdapterErrorObservationContract =
  DeepReadonly<Static<typeof externalAdapterErrorObservationSchema>>;

export const externalAdapterTelemetryObservationSchema = Type.Object(
  {
    name: Type.String({ minLength: 1 }),
    value: Type.Union([Type.Number(), Type.String(), Type.Boolean()]),
    unit: Type.Optional(Type.String({ minLength: 1 })),
  },
  { additionalProperties: false },
);

export type ExternalAdapterTelemetryObservationContract =
  DeepReadonly<Static<typeof externalAdapterTelemetryObservationSchema>>;

export const externalAdapterResponseV1Schema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.externalAdapterResponse),
    protocol_version: Type.Literal(externalAdapterProtocolVersion),
    invocation_id: Type.String({ minLength: 1 }),
    adapter_id: Type.String({ minLength: 1 }),
    status: stringEnum(externalAdapterStatuses),
    stdout: Type.Optional(Type.String()),
    stderr: Type.Optional(Type.String()),
    exit_code: Type.Optional(Type.Union([Type.Integer(), Type.Null()])),
    output: Type.Optional(unknownRecordSchema()),
    artifacts: Type.Optional(Type.Array(externalAdapterArtifactObservationSchema)),
    errors: Type.Optional(Type.Array(externalAdapterErrorObservationSchema)),
    telemetry: Type.Optional(Type.Array(externalAdapterTelemetryObservationSchema)),
    metadata: Type.Optional(unknownRecordSchema()),
    observed_at: dateTimeStringSchema(),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.externalAdapterResponse,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.externalAdapterResponse,
    additionalProperties: false,
  },
);

export type ExternalAdapterResponseContract =
  DeepReadonly<Static<typeof externalAdapterResponseV1Schema>>;

export const externalAdapterHostResolutionFrameV1Schema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.externalAdapterHostResolution),
    protocol_version: Type.Literal(externalAdapterProtocolVersion),
    frame_id: Type.String({ minLength: 1 }),
    invocation_id: Type.String({ minLength: 1 }),
    adapter_id: Type.String({ minLength: 1 }),
    request: resolutionRequestSchema,
    requested_at: dateTimeStringSchema(),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.externalAdapterHostResolution,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.externalAdapterHostResolution,
    additionalProperties: false,
  },
);

export type ExternalAdapterHostResolutionFrameContract =
  DeepReadonly<Static<typeof externalAdapterHostResolutionFrameV1Schema>>;

export const externalAdapterCancellationFrameV1Schema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.externalAdapterCancellation),
    protocol_version: Type.Literal(externalAdapterProtocolVersion),
    frame_id: Type.String({ minLength: 1 }),
    invocation_id: Type.String({ minLength: 1 }),
    adapter_id: Type.String({ minLength: 1 }),
    reason: Type.String({ minLength: 1 }),
    requested_at: dateTimeStringSchema(),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.externalAdapterCancellation,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.externalAdapterCancellation,
    additionalProperties: false,
  },
);

export type ExternalAdapterCancellationFrameContract =
  DeepReadonly<Static<typeof externalAdapterCancellationFrameV1Schema>>;

export function validateExternalAdapterManifestContract(
  value: unknown,
  label = "external_adapter_manifest",
): ExternalAdapterManifestContract {
  return validateContractSchema(externalAdapterManifestV1Schema, value, label);
}

export function validateExternalAdapterInvocationContract(
  value: unknown,
  label = "external_adapter_invocation",
): ExternalAdapterInvocationContract {
  return validateContractSchema(externalAdapterInvocationV1Schema, value, label);
}

export function validateExternalAdapterResponseContract(
  value: unknown,
  label = "external_adapter_response",
): ExternalAdapterResponseContract {
  return validateContractSchema(externalAdapterResponseV1Schema, value, label);
}

export function validateExternalAdapterHostResolutionFrameContract(
  value: unknown,
  label = "external_adapter_host_resolution",
): ExternalAdapterHostResolutionFrameContract {
  return validateContractSchema(externalAdapterHostResolutionFrameV1Schema, value, label);
}

export function validateExternalAdapterCancellationFrameContract(
  value: unknown,
  label = "external_adapter_cancellation",
): ExternalAdapterCancellationFrameContract {
  return validateContractSchema(externalAdapterCancellationFrameV1Schema, value, label);
}

export function validateExternalAdapterCredentialRequestContract(
  value: unknown,
  label = "external_adapter_credential_request",
): ExternalAdapterCredentialRequestContract {
  return validateContractSchema(externalAdapterCredentialRequestV1Schema, value, label);
}

File: crates/runx-contracts/src/external_adapter.rs
//! External adapter contract types.
use serde::{Deserialize, Deserializer, Serialize};

use crate::schema::{IsoDateTime, NonEmptyString, RunxSchema};
use crate::{JsonNumber, JsonObject, Reference, ResolutionRequest};

pub const EXTERNAL_ADAPTER_PROTOCOL_VERSION: &str = "runx.external_adapter.v1";

/// The const `protocol_version` discriminant shared by every external-adapter
/// frame.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum ExternalAdapterProtocolVersion {
    #[serde(rename = "runx.external_adapter.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum ExternalAdapterManifestSchema {
    #[serde(rename = "runx.external_adapter.manifest.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum ExternalAdapterCredentialRequestSchema {
    #[serde(rename = "runx.external_adapter.credential_request.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum ExternalAdapterInvocationSchema {
    #[serde(rename = "runx.external_adapter.invocation.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum ExternalAdapterHostResolutionSchema {
    #[serde(rename = "runx.external_adapter.host_resolution.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum ExternalAdapterCancellationSchema {
    #[serde(rename = "runx.external_adapter.cancellation.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum ExternalAdapterTransportKind {
    Process,
    Http,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum ExternalAdapterStatus {
    Completed,
    Failed,
    HostResolutionRequested,
    Cancelled,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum ExternalAdapterCredentialPurpose {
    ProviderApi,
    Registry,
    ArtifactStore,
    WebhookVerification,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalAdapterTransport {
    pub kind: ExternalAdapterTransportKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub command: Option<NonEmptyString>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub args: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub endpoint: Option<NonEmptyString>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalAdapterCredentialNeed {
    pub purpose: ExternalAdapterCredentialPurpose,
    pub provider: NonEmptyString,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scope_refs: Option<Vec<Reference>>,
    pub required: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalAdapterSandboxIntent {
    pub profile: NonEmptyString,
    pub network: bool,
    pub cwd_policy: NonEmptyString,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub writable_paths: Option<Vec<NonEmptyString>>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalAdapterTimeouts {
    pub startup_ms: u64,
    pub invocation_ms: u64,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_adapter.manifest.v1")]
pub struct ExternalAdapterManifest {
    pub schema: ExternalAdapterManifestSchema,
    pub protocol_version: ExternalAdapterProtocolVersion,
    pub adapter_id: NonEmptyString,
    pub name: NonEmptyString,
    pub version: NonEmptyString,
    pub supported_source_types: Vec<NonEmptyString>,
    pub transport: ExternalAdapterTransport,
    pub timeouts: ExternalAdapterTimeouts,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credential_needs: Option<Vec<ExternalAdapterCredentialNeed>>,
    pub sandbox_intent: ExternalAdapterSandboxIntent,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<JsonObject>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalAdapterCredentialReference {
    pub credential_ref: Reference,
    pub provider: NonEmptyString,
    pub purpose: ExternalAdapterCredentialPurpose,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_adapter.credential_request.v1")]
pub struct ExternalAdapterCredentialRequest {
    pub schema: ExternalAdapterCredentialRequestSchema,
    pub protocol_version: ExternalAdapterProtocolVersion,
    pub request_id: NonEmptyString,
    pub adapter_id: NonEmptyString,
    pub invocation_id: NonEmptyString,
    pub credential_refs: Vec<ExternalAdapterCredentialReference>,
    pub requested_at: IsoDateTime,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_adapter.invocation.v1")]
pub struct ExternalAdapterInvocation {
    pub schema: ExternalAdapterInvocationSchema,
    pub protocol_version: ExternalAdapterProtocolVersion,
    pub invocation_id: NonEmptyString,
    pub adapter_id: NonEmptyString,
    pub run_id: NonEmptyString,
    pub step_id: NonEmptyString,
    pub source_type: NonEmptyString,
    pub skill_ref: NonEmptyString,
    pub harness_ref: Reference,
    pub host_ref: Reference,
    pub inputs: JsonObject,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolved_inputs: Option<JsonObject>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<NonEmptyString>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub receipt_dir: Option<NonEmptyString>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env: Option<JsonObject>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credential_refs: Option<Vec<ExternalAdapterCredentialReference>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<JsonObject>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalAdapterArtifactObservation {
    pub artifact_ref: Reference,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalAdapterErrorObservation {
    pub code: String,
    pub message: String,
    pub retryable: bool,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, RunxSchema)]
#[serde(untagged)]
pub enum ExternalAdapterTelemetryValue {
    Number(JsonNumber),
    String(String),
    Bool(bool),
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalAdapterTelemetryObservation {
    pub name: String,
    pub value: ExternalAdapterTelemetryValue,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unit: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_adapter.response.v1")]
pub struct ExternalAdapterResponse {
    pub schema: String,
    pub protocol_version: String,
    pub invocation_id: String,
    pub adapter_id: String,
    pub status: ExternalAdapterStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stdout: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stderr: Option<String>,
    #[serde(
        default,
        deserialize_with = "deserialize_optional_nullable_i64",
        skip_serializing_if = "Option::is_none"
    )]
    pub exit_code: Option<Option<i64>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output: Option<JsonObject>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub artifacts: Option<Vec<ExternalAdapterArtifactObservation>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub errors: Option<Vec<ExternalAdapterErrorObservation>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub telemetry: Option<Vec<ExternalAdapterTelemetryObservation>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<JsonObject>,
    pub observed_at: String,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_adapter.host_resolution.v1")]
pub struct ExternalAdapterHostResolutionFrame {
    pub schema: ExternalAdapterHostResolutionSchema,
    pub protocol_version: ExternalAdapterProtocolVersion,
    pub frame_id: NonEmptyString,
    pub invocation_id: NonEmptyString,
    pub adapter_id: NonEmptyString,
    pub request: ResolutionRequest,
    pub requested_at: IsoDateTime,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_adapter.cancellation.v1")]
pub struct ExternalAdapterCancellationFrame {
    pub schema: ExternalAdapterCancellationSchema,
    pub protocol_version: ExternalAdapterProtocolVersion,
    pub frame_id: NonEmptyString,
    pub invocation_id: NonEmptyString,
    pub adapter_id: NonEmptyString,
    pub reason: NonEmptyString,
    pub requested_at: IsoDateTime,
}

fn deserialize_optional_nullable_i64<'de, D>(
    deserializer: D,
) -> Result<Option<Option<i64>>, D::Error>
where
    D: Deserializer<'de>,
{
    Option::<i64>::deserialize(deserializer).map(Some)
}
